lightning/src/routing/network_graph.rs

```diff
@@ -103,6 +103,66 @@ impl_writeable_tlv_based_enum!(NetworkUpdate,
 	},
 ;);
 
+/// An [`EventHandler`] decorator for applying updates from [`Event::PaymentFailed`] to the
+/// [`NetworkGraph`].
+pub struct NetworkUpdateHandler<G: Deref<Target=NetworkGraph>, L: Deref, E: EventHandler>
+where L::Target: Logger {
+	secp_ctx: Secp256k1<secp256k1::VerifyOnly>,
+	network_graph: G,
+	logger: L,
+	inner: E,
+}
+
+impl<G: Deref<Target=NetworkGraph>, L: Deref, E: EventHandler> NetworkUpdateHandler<G, L, E>
+where L::Target: Logger {
+	/// Creates a handler that decorates `event_handler` with functionality for updating the given
+	/// `network_graph`.
+	pub fn new(network_graph: G, logger: L, event_handler: E) -> Self {
+		Self {
+			secp_ctx: Secp256k1::verification_only(),
+			network_graph,
+			logger,
+			inner: event_handler,
+		}
+	}
+
+	/// Applies changes to the [`NetworkGraph`] from the given update.
+	fn handle_network_update(&self, update: &NetworkUpdate) {
+		match *update {
+			NetworkUpdate::ChannelUpdateMessage { ref msg } => {
+				let short_channel_id = msg.contents.short_channel_id;
+				let is_enabled = msg.contents.flags & (1 << 1) != (1 << 1);
+				let status = if is_enabled { "enabled" } else { "disabled" };
+				log_debug!(self.logger, "Updating channel with channel_update from a payment failure. Channel {} is {}.", short_channel_id, status);
+				let _ = self.network_graph.update_channel(msg, &self.secp_ctx);
+			},
+			NetworkUpdate::ChannelClosed { short_channel_id, is_permanent } => {
+				let action = if is_permanent { "Removing" } else { "Disabling" };
+				log_debug!(self.logger, "{} channel graph entry for {} due to a payment failure.", action, short_channel_id);
+				self.network_graph.close_channel_from_update(short_channel_id, is_permanent);
+			},
+			NetworkUpdate::NodeFailure { ref node_id, is_permanent } => {
+				let action = if is_permanent { "Removing" } else { "Disabling" };
+				log_debug!(self.logger, "{} node graph entry for {} due to a payment failure.", action, node_id);
+				self.network_graph.fail_node(node_id, is_permanent);
+			},
+		}
+	}
+}
+
+impl<G: Deref<Target=NetworkGraph>, L: Deref, E: EventHandler> EventHandler for NetworkUpdateHandler<G, L, E>
+where L::Target: Logger {
+	fn handle_event(&self, event: Event) {
+		if let Event::PaymentFailed { payment_hash: _, rejected_by_dest: _, network_update, .. } = &event {
+			if let Some(network_update) = network_update {
+				self.handle_network_update(network_update);
+			}
+		}
+
+		self.inner.handle_event(event)
```

Comment:
Hmmmmmm, have we considered maybe just allowing events to be handled by multiple handlers? Passing them off to an inner handler works, I suppose, but its kinda awkward. I guess I'm not a huge fan of passing the events by reference instead of value, but most of them have no heap allocation, and eg `SpendableOutputs` should be pretty rare (requires on-chain txn).

Reply:
Hmm, yea, good point. This isn't a trait change, though, so we could do different things for different handlers, right?

Reply:
Not quite sure I follow what you mean by do "different things for different handlers". Do you have a concrete example that may clarify?

Reply:
As in, there's nothing here that requires that we handle ignoring/passing down events the same for NetworkGraph as we do for PaymentRetryer, or whatever, is there?

Reply:
Aside from, of course, the fact that EventHandler takes the event by move, not by reference, but we could change that with minimal overhead, I think.

Reply:
Correct. Though you're right that passing by move is not as flexible since it doesn't allow passing to multiple event handlers in interesting ways. Or post-handling an event after passing to an inner handler for that matter. Updated to pass by reference.

Reply:
Haven't pushed it yet, but I went with implementing it on `NetGraphMsgHandler`. It's also nice to not have to know about a decorated `EventHandler` when deserializing a `NetworkGraph`.

That said, just want to point out a drawback with either is that the [guide](https://lightningdevkit.org/docs/build_node_rust/) for setting up a node will need to change. Now event handling (step 16) can no longer be discussed independently of initializing `NetGraphMsgHandler` (optional step 12).

cc: @valentinewallace Any opinions on this?

Reply:
Thanks for pointing that out. That is a slight drawback but not a deal breaker IMO

Reply:
I think that's ok. Indeed, ideally we'd have everything be nice and separate, but if we're gonna have something related to the network graph handle events, that's the outcome we're gonna have. It is nice that all the bindings have moved towards using the lightning-background-processor crate, though, so it'll all be common infrastructure in any case.